bpf/flows.c

```diff
@@ -65,89 +116,177 @@ static inline int fill_iphdr(struct iphdr *ip, void *data_end, struct flow *flow
     case IPPROTO_TCP: {
         struct tcphdr *tcp = (void *)ip + sizeof(*ip);
         if ((void *)tcp + sizeof(*tcp) <= data_end) {
-            flow->transport.src_port = __bpf_ntohs(tcp->source);
-            flow->transport.dst_port = __bpf_ntohs(tcp->dest);
+            flow_id->src_port = __bpf_ntohs(tcp->source);
+            flow_id->dst_port = __bpf_ntohs(tcp->dest);
+             if (tcp->fin) {
+                 *flags= *flags | TCP_FIN_FLAG;
+             }
+             if (tcp->rst) {
+                 *flags= *flags | TCP_RST_FLAG;
+             }
         }
     } break;
     case IPPROTO_UDP: {
         struct udphdr *udp = (void *)ip + sizeof(*ip);
         if ((void *)udp + sizeof(*udp) <= data_end) {
-            flow->transport.src_port = __bpf_ntohs(udp->source);
-            flow->transport.dst_port = __bpf_ntohs(udp->dest);
+            flow_id->src_port = __bpf_ntohs(udp->source);
+            flow_id->dst_port = __bpf_ntohs(udp->dest);
         }
     } break;
     default:
         break;
     }
     return SUBMIT;
 }
-// sets flow fields from Ethernet header information
-static inline int fill_ethhdr(struct ethhdr *eth, void *data_end, struct flow *flow) {
+// // sets flow fields from Ethernet header information
+static inline int fill_ethhdr(struct ethhdr *eth, void *data_end, flow_id_v *flow_id, u32 *flags) {
     if ((void *)eth + sizeof(*eth) > data_end) {
         return DISCARD;
     }
-    __builtin_memcpy(flow->data_link.dst_mac, eth->h_dest, ETH_ALEN);
-    __builtin_memcpy(flow->data_link.src_mac, eth->h_source, ETH_ALEN);
-    flow->protocol = __bpf_ntohs(eth->h_proto);
-    // TODO: ETH_P_IPV6
-    if (flow->protocol == ETH_P_IP) {
+    __builtin_memcpy(flow_id->dst_mac, eth->h_dest, ETH_ALEN);
+    __builtin_memcpy(flow_id->src_mac, eth->h_source, ETH_ALEN);
+    flow_id->eth_protocol = __bpf_ntohs(eth->h_proto);
+
+    if (flow_id->eth_protocol == ETH_P_IP) {
         struct iphdr *ip = (void *)eth + sizeof(*eth);
-        return fill_iphdr(ip, data_end, flow);
-    } else if (flow->protocol == ETH_P_IPV6) {
+        return fill_iphdr(ip, data_end, flow_id, flags);
+    } else if (flow_id->eth_protocol == ETH_P_IPV6) {
         struct ipv6hdr *ip6 = (void *)eth + sizeof(*eth);
-        return fill_ip6hdr(ip6, data_end, flow);
+        return fill_ip6hdr(ip6, data_end, flow_id, flags);
+    } else {
+        // TODO : Need to implement other specific ethertypes if needed
+        // For now other parts of flow id remain zero
+        memset (&(flow_id->src_ip),0, sizeof(struct in6_addr));
+        memset (&(flow_id->dst_ip),0, sizeof(struct in6_addr));
+        flow_id->protocol = 0;
+        flow_id->src_port = 0;
+        flow_id->dst_port = 0;
     }
     return SUBMIT;
 }
 
-// parses flow information for a given direction (ingress/egress)
-static inline int flow_parse(struct __sk_buff *skb, u8 direction) {
+static inline void export_flow_id (flow *my_flow_key, flow_id_v my_flow_id, u8 direction) {
```

Comment:
for clarity and brevity, I'd rename values as:
- `my_flow_key` --> `key`
- `my_flow_id` --> `id`